dask/dataframe/io/parquet/fastparquet.py

```diff
@@ -12,23 +13,57 @@
 
 try:
     import fastparquet
-    from fastparquet import ParquetFile
+    from fastparquet import ParquetFile as FPParquetFile
     from fastparquet.util import get_file_scheme
     from fastparquet.util import ex_from_sep, val_to_num, groupby_types
     from fastparquet.writer import partition_on_columns, make_part_file
 except ImportError:
     pass
 
-from .utils import _parse_pandas_metadata, _normalize_index_columns, _analyze_paths
+from .utils import (
+    _parse_pandas_metadata,
+    _normalize_index_columns,
+    _analyze_paths,
+    _flatten_filters,
+    _row_groups_to_parts,
+)
 from ..utils import _meta_from_dtypes
 from ...utils import UNKNOWN_CATEGORIES
+from ...methods import concat
+
+try:
+    # Required for distrubuted
+    import pickle
+except ImportError:
+    pickle = None
+
 
 #########################
 # Fastparquet interface #
 #########################
 from .utils import Engine
 
 
+class ParquetFile(FPParquetFile):
+    # Define our own `ParquetFile` as a subclass
+    # of `fastparquet.ParquetFile`.  This allows
+    # us to strip the `key_value_metadata` and
+    # `fmd` attributes before sending to the
+    # workers.
+
+    @property
+    def pandas_metadata(self):
+        if hasattr(self, "_pandas_metadata"):
+            return self._pandas_metadata
+        return super().pandas_metadata
+
+    @property
+    def has_pandas_metadata(self):
+        if hasattr(self, "_has_pandas_metadata"):
+            return self._has_pandas_metadata
+        return super().has_pandas_metadata
```

Comment:
@martindurant - Note that I needed to override `ParquetFile` to make it possible to remove `fmd` and `key_value_metadata`.

Reply:
Hmm - But now it looks like the `fmd` removal is running into a problem in `__setstate__` for [fastparquet#549](https://github.com/dask/fastparquet/pull/549)

```
~/workspace/cudf-0.18/distributed/distributed/protocol/pickle.py in loads()
     73             return pickle.loads(x, buffers=buffers)
     74         else:
---> 75             return pickle.loads(x)
     76     except Exception as e:
     77         logger.info("Failed to deserialize %s", x[:10000], exc_info=True)

~/workspace/cudf-0.18/fastparquet/fastparquet/api.py in __setstate__()
    545     def __setstate__(self, state):
    546         self.__dict__.update(state)
--> 547         self._set_attrs()
    548 
    549     def __str__(self):

~/workspace/cudf-0.18/fastparquet/fastparquet/api.py in _set_attrs()
    142     def _set_attrs(self):
    143         fmd = self.fmd
--> 144         self.version = fmd.version
    145         self._schema = fmd.schema
    146         self.row_groups = fmd.row_groups or []

AttributeError: 'NoneType' object has no attribute 'version'
```

Reply:
We can make more of the attributes into properties, but I thought that the getstate was enough for what you needed.

Why would you strip the fmd out? Without that, the ParquetFile class is pretty much useless.

Reply:
> We can make more of the attributes into properties

I mean upstream in fastparquet, which I would be happy to release whenever to be ready for this PR.

Reply:
> Why would you strip the fmd out? Without that, the ParquetFile class is pretty much useless.

Right now, on the read side, we only need the `pandas_metadata` and the `row_group`.  However, it may not be a huge size reduction to keep `fmd`, I'll investigate.

Reply:
Hm, you might have a point, if we've already parsed the pandas part, and determined dtypes anyway. If you can get away with it, without re-parsing the target file's metadata footer, sure! It would certainly make for a very small binary blob - but indeed it becomes a question of exactly what the instance *is* doing for us!

I may be getting a little confused at this point, happy to chat live, if you think it worthwhile.

Reply:
Also, if it's convenience, happy to make a classmethod on ParquetFile to construct instances out of whatever minimum information is passed into the read tasks.

Reply:
Okay - Im removing the `ParquetFile` override since leaving in `key_value_metadata` and `fmd` doesn't seem to be a problem.